src/api/web3.ts

```diff
@@ -0,0 +1,210 @@
+import Web3 from 'web3'
+import { Block } from 'web3-eth'
+import { provider } from 'web3-providers'
+import { TransactionReceipt, Transaction } from 'web3-core'
+
+import { windowLoaded } from './utils'
+
+declare global {
+    interface Window {
+        ethereum: provider & {
+            enable: () => Promise<string[]>
+            on?: () => EventListener
+        }
+        web3: Web3
+    }
+}
+
+export interface Web3FrontendAPI {
+    web3: Web3
+    getCurrentAccount: () => Promise<string>
+    getAccounts: () => Promise<string[]>
+    getBalance: (account?: string) => Promise<string>
+    getBlock: (blockNumberOrString: string | number) => Promise<Partial<Block>>
+    getNetwork: () => Promise<string | number>
+    getNetworkName: (id?: number | string) => Promise<string>
+    getNetworkProviderURL: (id: number | string) => string
+    getTransaction: (txHash: string) => Promise<Transaction>
+    getTransactionReceipt: (txHash: string) => Promise<TransactionReceipt>
+    // Utils
+    isAddress: (address: string) => boolean
+    isReadOnly: () => boolean
+    // Provider Specific
+    resetProvider(): void
+    setProvider(provider: string): void
+}
+
+export interface Web3SetupOptions {
+    customProvider?: provider | string
+    reloadOnAccountsChange?: boolean
+}
+
+let web3API: Web3FrontendAPI
+
+let readOnly: boolean
+
+export const getWeb3API = async (options: Web3SetupOptions = {}, force?: boolean): Promise<Web3FrontendAPI> => {
+    if (!force && web3API) return web3API
+
+    return setupWeb3(options)
+}
+
+const getProvider = async ({ customProvider, reloadOnAccountsChange }: Web3SetupOptions): Promise<Web3> => {
```

Comment:
`getWeb3` looks like a better name, since it's what it returns.
Note that there's another type, that is call `provider` and is a different thing.

Also, I would call the `customProvider` just `provider` for simplicity. We can see already in the interface the it is optional.